docs(internals): add sandbox mode evaluation (#26)
dgowrie/claude-workflows: internals/sandbox-mode-evaluation.md

> # Claude Code Sandbox Mode Evaluation

Author: 💡 No `README.md` entry for this new internals doc. Per repo conventions, new internals docs should get an entry in the README's internals section before merge.

Author: :zap: a35979f

How sandbox mode works, what it restricts, and when to use it vs. Docker isolation.

---

## Enabling Sandbox Mode

### Via settings.json (persistent, schema can change)

Use `/sandbox` once, then inspect settings at both scopes:

- `~/.claude/settings.json` (global defaults)
- `<repo>/.claude/settings.json` (project overrides, when present)

This doc intentionally avoids pinning an unverified `sandbox.enabled`/`sandbox.mode` schema example.
Key names and structure can vary across Claude Code versions.

### Via `/sandbox` command (interactive, ad-hoc)

Opens a menu to choose between auto-allow and regular permissions. Can toggle mid-session.

### Platform requirements

- **macOS**: works out of the box (Seatbelt)
- **Linux/WSL2**: install `bubblewrap` and `socat`
- **WSL1**: not supported

---

## What the Sandbox Restricts

### Filesystem

- Writes confined to CWD and subdirectories by default

> - Reads allowed broadly

Author: 💡 "Broadly" is vague — the sandbox read deny-list is essentially empty by default. "Unrestricted" is more precise.

```suggestion
- Reads are unrestricted by default
```

Author: :zap: a35979f

### Protected paths (different protection semantics)

- **Repo metadata/config paths (never auto-approved):** `.git`, `.claude`, `.vscode`, `.idea`, `.husky`
- **User-home config/credential paths (high-friction; typically prompt/deny):** `~/.gitconfig`, `~/.bashrc`, `~/.zshrc`, `~/.ssh`
- **System paths (outside normal write scope; prompt/deny):** `/bin`, `/usr`

### Enforcement

OS-level (Seatbelt on macOS, bubblewrap on Linux). All child processes inherit restrictions.

---

## Auto-Allow Mode Security Boundaries

### Auto-allows (no prompts)

- Bash commands within sandbox boundaries
- File writes/deletes inside CWD
- Network requests to pre-approved domains

### Still prompts

- Remote-modifying commands (`git push`, `gh pr create`, `gh api` POST/DELETE)
- Writes outside CWD (unless explicitly allowed)
- Network to unapproved domains

---

## Risk Assessment (repo-scoped session)

| Concern | Risk | Notes |
| --- | --- | --- |
| Deletes files in repo | Possible | Reversible via git |
| Modifies files outside repo | Blocked | Sandbox boundary |
| Pushes to remote | Prompts | Not auto-allowed |
| Creates/closes PRs or issues | Prompts | Not auto-allowed |

> | Exfiltrates code | Mitigated | Unapproved domains are blocked, but approved domains can still be exfiltration paths |

Author: 🟡 This covers the network-domain vector but misses tool-use side channels. The sandbox doesn't filter data embedded in tool call bodies — e.g., code exfiltrated via a `gh api` POST body to a legitimate endpoint, or sensitive content in commit messages pushed to a public repo. Network domain filtering is necessary but not sufficient.

```suggestion
| Exfiltrates code | Mitigated | Unapproved domains are blocked, but approved domains and tool-use side channels (e.g., data in API call bodies) are not filtered |
```

Author: :zap: a35979f

| Modifies git/shell config | Blocked | Protected path |
| Malicious subprocess escapes | Blocked | OS-enforced |

---

## Native Sandbox vs Docker

| Feature | Native Sandbox | Docker Container |
| --- | --- | --- |
| Setup | Built-in, zero deps (macOS) | Docker daemon + image management |
| Performance | Near-native | Higher overhead, slower I/O |
| Filesystem | Selective read/write | All-or-nothing volume mounts |
| Network | Domain-level filtering | Complete isolation unless ports exposed |
| Dev workflow | IDE, shell, tools seamless | Must enter container; separate tooling |
| Isolation strength | OS sandbox primitives | Full kernel-level isolation (stronger) |

> | Escape difficulty | `dangerouslyDisableSandbox` flag (prompts user) | Harder (unless `--privileged`) |

Author: 🟡 "prompts user" is accurate for interactive sessions but may not hold in headless/CI contexts depending on permission mode.

```suggestion
| Escape difficulty | `dangerouslyDisableSandbox` flag (prompts user in interactive sessions; headless behavior depends on permission mode) | Harder (unless `--privileged`) |
```

Author: :zap: a35979f

# Claude Code Sandbox Mode Evaluation

How sandbox mode works, what it restricts, and when to use it vs. Docker isolation.

---

## Enabling Sandbox Mode

### Via settings.json (persistent, silent)

```json
{
  "sandbox": {

>     "enabled": true,

Author: 🟡 The `sandbox.enabled` + `sandbox.mode` settings structure shown here may not reflect the actual Claude Code `settings.json` schema. If the shape is wrong, readers who copy-paste this config will get silent no-ops.

Worth verifying against a real session or the Claude Code source before merging.

Author: @copilot verify

Author: :zap: e22ded1

    "mode": "auto-allow"
  }
}
```

No per-session prompts. Every session starts sandboxed automatically.

### Via `/sandbox` command (interactive, ad-hoc)

Opens a menu to choose between auto-allow and regular permissions. Can toggle mid-session.

### Platform requirements

- **macOS**: works out of the box (Seatbelt)
- **Linux/WSL2**: install `bubblewrap` and `socat`
- **WSL1**: not supported

---

## What the Sandbox Restricts

### Filesystem

- Writes confined to CWD and subdirectories by default
- Reads allowed broadly

> - Customizable via `sandbox.filesystem.allowWrite`

Author: 🟡 Same concern — `sandbox.filesystem.allowWrite` may not be the actual settings key. The sandbox prompt shown at session start references `permissions.allow` patterns, not a nested `sandbox.filesystem` namespace. Spot-check against a real session's sandbox output.

Author: :zap: e22ded1

### Network

- Proxy-based domain filtering; only approved domains reachable

> - New domains trigger prompts (or auto-block with `allowManagedDomainsOnly`)

Author: 🟡 "Proxy-based domain filtering" and `allowManagedDomainsOnly` are plausible but unverified. The PR's own test plan flags "spot-check claims against current Claude Code sandbox behavior" as a TODO — these specific mechanism names and config keys are the highest-risk items for that check.

Author: :zap: e22ded1

- Domain-level only, no TLS inspection

### Protected paths (never auto-approved)

> `.git`, `.claude`, `.vscode`, `.idea`, `.husky`, `~/.gitconfig`, `~/.bashrc`, `~/.zshrc`, `~/.ssh`, system directories (`/bin`, `/usr`)

Author: 💡 This list mixes repo-local paths (`.git`, `.claude`), user-home paths (`~/.gitconfig`, `~/.ssh`), and system paths (`/bin`). These have different protection semantics (never writable vs. always prompt). Separating into sub-categories would make the doc more precise.

Author: @copilot handle this per the recommendation

Author: :zap: 6c38195

### Enforcement

OS-level (Seatbelt on macOS, bubblewrap on Linux). All child processes inherit restrictions.

---

## Auto-Allow Mode Security Boundaries

### Auto-allows (no prompts)

- Bash commands within sandbox boundaries
- File writes/deletes inside CWD
- Network requests to pre-approved domains

### Still prompts

- Remote-modifying commands (`git push`, `gh pr create`, `gh api` POST/DELETE)
- Writes outside CWD (unless explicitly allowed)
- Network to unapproved domains

---

## Risk Assessment (repo-scoped session)

| Concern | Risk | Notes |
| --- | --- | --- |
| Deletes files in repo | Possible | Reversible via git |
| Modifies files outside repo | Blocked | Sandbox boundary |
| Pushes to remote | Prompts | Not auto-allowed |
| Creates/closes PRs or issues | Prompts | Not auto-allowed |

> | Exfiltrates code | Blocked | Network proxy |

Author: 💡 "Blocked" is a strong claim here. Sandbox blocks unapproved domains, but approved domains (npm registry, GitHub API) could still be exfiltration vectors. Consider softening to "Mitigated" with a note about the approved-domain caveat.

Author: @copilot make this change

Author: :zap: b02b189
